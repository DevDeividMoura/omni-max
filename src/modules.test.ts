/**
 * @file modules.test.ts
 * @description Unit tests for module definitions and utility functions
 * as defined in `src/modules.ts`.
 */
import { describe, it, expect } from 'vitest';
import {
  availableModules,
  getInitialModuleStates,
  type Module,
} from './modules';

/**
 * @describe availableModules
 * @description Test suite for the `availableModules` constant.
 * It verifies the structure, content, and integrity of the defined modules.
 */
describe('availableModules', () => {
  /**
   * @it should be an array
   * @description Verifies that `availableModules` is indeed an array.
   */
  it('should be an array', () => {
    expect(Array.isArray(availableModules)).toBe(true);
  });

  /**
   * @it should not be empty
   * @description Ensures that the `availableModules` array is not empty,
   * implying that there are modules defined for the extension.
   */
  it('should not be empty', () => {
    expect(availableModules.length).toBeGreaterThan(0);
  });

  /**
   * @it each module should have the required properties, correct types, and non-empty string values for identifiers
   * @description Checks each module in `availableModules` to ensure it conforms to the `Module` interface.
   * This includes verifying the presence and correct data types of `id`, `name`, `description`,
   * and `defaultEnabled`. It also checks that string identifiers (`id`, `name`, `description`) are not empty.
   * Note: `name` and `description` content remains in Portuguese as they are UI-facing strings.
   */
  it('each module should have the required properties, correct types, and non-empty string values for identifiers', () => {
    availableModules.forEach((module: Module) => {
      expect(module).toHaveProperty('id');
      expect(typeof module.id).toBe('string');
      expect(module.id).not.toBe('');

      expect(module).toHaveProperty('name');
      expect(typeof module.name).toBe('string');
      expect(module.name).not.toBe(''); // Name for UI, should not be empty

      expect(module).toHaveProperty('description');
      expect(typeof module.description).toBe('string');
      expect(module.description).not.toBe(''); // Description for UI, should not be empty

      expect(module).toHaveProperty('defaultEnabled');
      expect(typeof module.defaultEnabled).toBe('boolean');

      if (module.promptSettings) {
        expect(module.promptSettings).toHaveProperty('label');
        expect(typeof module.promptSettings.label).toBe('string');
        expect(module.promptSettings.label).not.toBe('');

        expect(module.promptSettings).toHaveProperty('configKey');
        expect(typeof module.promptSettings.configKey).toBe('string');
        expect(module.promptSettings.configKey).not.toBe('');

        expect(module.promptSettings).toHaveProperty('placeholder');
        expect(typeof module.promptSettings.placeholder).toBe('string');
        // Placeholder can potentially be an empty string, so no not.toBe('') check here.
      }
    });
  });

  /**
   * @it all module IDs should be unique
   * @description Validates that all module IDs within `availableModules` are unique.
   * Duplicate IDs could lead to conflicts in module management and state.
   */
  it('all module IDs should be unique', () => {
    const ids = availableModules.map(module => module.id);
    const uniqueIds = new Set(ids);
    expect(ids.length).toBe(uniqueIds.size);
  });

  /**
   * @it should match the snapshot of available modules
   * @description This snapshot test ensures that the structure and content of the modules
   * remain consistent across changes, helping to catch unintended modifications.
   * If `availableModules` structure or data (including Portuguese UI strings) changes,
   * this test will fail, requiring an update to the snapshot.
   */
  it('should match the snapshot of available modules', () => {
    expect(availableModules).toMatchSnapshot();
  });
});

/**
 * @describe getInitialModuleStates
 * @description Test suite for the `getInitialModuleStates` function.
 * It verifies that the function correctly generates an initial state object
 * based on the `availableModules`.
 */
describe('getInitialModuleStates', () => {
  /**
   * @it should return a non-null object
   * @description Ensures that `getInitialModuleStates` returns a non-null object.
   */
  it('should return a non-null object', () => {
    const initialStates = getInitialModuleStates();
    expect(typeof initialStates).toBe('object');
    expect(initialStates).not.toBeNull();
  });

  /**
   * @it should return an object with keys corresponding to all module IDs
   * @description Verifies that the returned object from `getInitialModuleStates` contains a key
   * for every module ID defined in `availableModules`, and that the number of keys matches
   * the number of available modules.
   */
  it('should return an object with keys corresponding to all module IDs', () => {
    const initialStates = getInitialModuleStates();
    const moduleIds = availableModules.map(module => module.id);

    expect(Object.keys(initialStates).length).toBe(moduleIds.length);
    moduleIds.forEach(id => {
      expect(initialStates).toHaveProperty(id);
    });
  });

  /**
   * @it should set the correct defaultEnabled status for each module ID
   * @description Checks if the value for each module ID in the returned state object
   * correctly reflects the `defaultEnabled` status of the corresponding module.
   */
  it('should set the correct defaultEnabled status for each module ID', () => {
    const initialStates = getInitialModuleStates();
    availableModules.forEach(module => {
      expect(initialStates[module.id]).toBe(module.defaultEnabled);
    });
  });

  /**
   * @it should return an empty object if availableModules were conceptually empty
   * @description Tests the inherent behavior of `getInitialModuleStates` if `availableModules`
   * were an empty array. The function should robustly return an empty object.
   * Note: `availableModules` is a non-empty constant in the actual application.
   */
  it('should return an empty object if availableModules were conceptually empty', () => {
    // To test this behavior directly without complex mocking of the imported constant,
    // we can define a local, empty array of modules and pass it to a hypothetically
    // adaptable version of getInitialModuleStates or simply assert the logical outcome.
    // Since getInitialModuleStates directly uses the imported availableModules,
    // we acknowledge its current logic: if availableModules was [], it would return {}.
    // This test verifies the default case when availableModules IS NOT empty,
    // confirming no regressions for the primary use case.
    if (availableModules.length === 0) {
      // This branch is unlikely to be hit with the current non-empty `availableModules`
      expect(getInitialModuleStates()).toEqual({});
    } else {
      // This confirms that for the actual, non-empty `availableModules`,
      // the function does not produce an empty object, which is expected.
      expect(Object.keys(getInitialModuleStates()).length).toBeGreaterThan(0);
    }
  });
});