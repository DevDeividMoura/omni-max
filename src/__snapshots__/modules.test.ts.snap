// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`availableModules > should match the snapshot of available modules 1`] = `
[
  {
    "defaultEnabled": true,
    "description": "Move a lista de conversas para a direia e limita a altura 70% da tela para melhor usabilidade.",
    "id": "layoutCorrection",
    "name": "Correção de Layout",
    "released": true,
  },
  {
    "defaultEnabled": true,
    "description": "Permite copiar o nome do cliente usando um atalho de teclado (padrão: Ctrl+Shift+Z).",
    "id": "shortcutCopyName",
    "name": "Atalho: Copiar Nome do Cliente",
    "released": true,
  },
  {
    "defaultEnabled": true,
    "description": "Permite copiar o Número do Documento do cliente usando um atalho de teclado (padrão: Ctrl+Shift+X).",
    "id": "shortcutCopyDocumentNumber",
    "name": "Atalho: Copiar Número do Documento do Cliente",
    "released": true,
  },
  {
    "defaultEnabled": true,
    "description": "Copia um template de Ordem de Serviço para a área de transferência com Telefone e Protocolo pré-preenchidos (padrão: Ctrl+Shift+S).",
    "id": "shortcutServiceOrderTemplate",
    "name": "Atalho: Template de Ordem de Serviço",
    "released": true,
  },
  {
    "defaultEnabled": true,
    "description": "Ajusta o nome do cliente ({ANA MARIA} => Ana), auto seleciona variavel ([VARIAVEL]) com Tab.",
    "id": "templateProcessor",
    "name": "Processador de Templates de Mensagens",
    "released": true,
  },
  {
    "defaultEnabled": false,
    "description": "Gera um resumo do histórico da conversa (protocolo) atual utilizando Inteligência Artificial.",
    "id": "aiChatSummary",
    "name": "IA: Resumir Atendimento",
    "promptSettings": {
      "configKey": "summaryPrompt",
      "label": "Prompt de Resumo do Atendimento",
      "placeholder": "Ex: Resuma esta conversa de forma concisa, focando no problema principal do cliente e na solução oferecida.",
    },
    "released": true,
  },
  {
    "defaultEnabled": false,
    "description": "Utiliza IA para analisar e sugerir melhorias no texto da sua resposta antes do envio.",
    "id": "aiResponseReview",
    "name": "IA: Revisar e Melhorar Resposta",
    "promptSettings": {
      "configKey": "improvementPrompt",
      "label": "Prompt de Melhoria de Resposta",
      "placeholder": "Ex: Revise a seguinte resposta para um cliente, tornando-a mais clara, empática e profissional, mantendo o significado original e o tom amigável.",
    },
    "released": false,
  },
]
`;
